### DISCUSSION TITLE
Macros for serialization?

### DISCUSSION THREAD

**Discussion by ryan-minato** [PyO3/pyo3#3957] [Questions]
I have a use case for a distributed system (well, it's ray data), where all objects need to be serializable using pickle. So I have to write a lot of similar code #100 , and my IDE keeps complaining about too much repetition. I wonder if there's a better way.

like this([url](https://github.com/PyO3/pyo3/issues/100#issuecomment-1219631108))
``` Python
pub fn __setstate__(&mut self, state: Vec<u8>) -> PyResult<()> {
    *self = deserialize(&state).unwrap();
    Ok(())
}
pub fn __getstate__(&self) -> PyResult<Vec<u8>> {
    Ok(serialize(&self).unwrap())
}
pub fn __getnewargs__(&self) -> PyResult<(f64, f64)> {
    Ok((self.my_first_arg, self.my_second_arg))
}
```

I searched the documentation and issues and found that there doesn't seem to be a derive macro (or other method) to reuse this logic. I'm a Rust newbie, but I think a derive macro might be a good ideal. Does pyo3 have any plans to add it?

**Comment by davidhewitt** [comment]
I think this is tracked in #100. It's not been a priority for me to go further right now, though I think something like #1375 would naturally fit with pickle support.
